File: sourceFile/Java/15_ThreeSum.java
/*
 * @Author: AlanGolphi
 * @Date: 2020-07-20 16:05:46
 * @LastEditTime: 2020-07-20 16:28:41
 */ 
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new LinkedList<>();
        for (int i = 0; i < nums.length-2; i++) {
            if (i == 0 || (i > 0 && nums[i] != nums[i-1])) {
                int lo = i+1, hi = nums.length-1, sum = 0 - nums[i];
                while (lo < hi) {
                    if (nums[lo] + nums[hi] == nums[i]) {
                        res.add(Arrays.asList(nums[i], nums[lo], nums[hi]));
                        while (lo < hi && nums[lo] == nums[lo+1]) lo++;
                        while (lo < hi && nums[hi] == nums[hi-1]) hi--;
                        lo++; hi--;
                    } else if (nums[lo] + nums[hi] < sum) lo++;
                    else hi--;
                }
            }
        }
        return res;
    }
}
// 1.排序
// 2.固定 i ，并在 i 后面的数的两端设置双指针进行遍历
// 3.如果两数之和小了，左指针右移；如果大了，右指针左移

File: sourceFile/Js/15_ThreeSum.js
/*
 * @Author: AlanGolphi
 * @Date: 2021-03-29 18:05:48
 * @LastEditTime: 2021-04-21 19:50:50
 */
// 给一个数组，查找数组中是否存在三个元素相加等于0，存在的话就组成一个数组，最后返回数组的数组

//1.排序 nums.sort((a, b) => a - b) 升序排序
//2. for() 循环里定义 left right指针，通过比较 nums[left] + nums[right] 和 target，移动 left 和 right；

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
function threeSum(nums) {
    let res = [];
    nums.sort((a,b) => a - b);//优雅的升序排序算法
    for (let i = 0, len = nums.length; i < len; i++) {
        if (nums[i] == nums[i - 1]) {
            continue;//保证最后结果不会输出重复项
        }
        let target = -nums[i];//求三数之和=n，则target = n - nums[i];
        let left = i + 1;
        let right = len - 1;
        while (left < right) {
            if (nums[left] + nums[right] == target) {
                res.push([-target, nums[left], nums[right]]);
                while (nums[left] == nums[left + 1]){
                    left ++;
                }
                while (nums[right] == nums[right - 1]) {
                    right --;
                }
                left++;
                right--;
            }
            else if (nums[left] + nums[right] < target) {
                left++;
            } else if (nums[left] + nums[right] > target) {
                right--;
            }
        }
        
    }
    return res;
}